File: chrome_app/credentials.js
// Copyright 2014 Mike Tsao <mike@sowbug.com>

// Permission is hereby granted, free of charge, to any person
// obtaining a copy of this software and associated documentation
// files (the "Software"), to deal in the Software without
// restriction, including without limitation the rights to use, copy,
// modify, merge, publish, distribute, sublicense, and/or sell copies
// of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
// BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

'use strict';

// The Credentials model keeps track of secrets that lock/unlock and
// encrypt/decrypt things.

/**
 * @constructor
 */
function Credentials() {
  this.init = function() {
    this.check = undefined;
    this.ephemeralKeyEncrypted = undefined;
    this.salt = undefined;

    this.isLocked = true;
  };
  this.init();

  this.toStorableObject = function() {
    var o = {};
    o["check"] = this.check;
    o["ekey_enc"] = this.ephemeralKeyEncrypted;
    o["salt"] = this.salt;
    return o;
  };

  this.loadStorableObject = function(o) {
    return new Promise(function(resolve, reject) {
      this.init();
      this.check = o["check"];
      this.ephemeralKeyEncrypted = o["ekey_enc"];
      this.salt = o["salt"];
      this.loadCredentials().then(resolve);
    }.bind(this));
  };

  this.isPassphraseSet = function() {
    return !!this.check;
  };

  // TODO(miket): rename to just isLocked (& avoid collision)
  this.isWalletLocked = function() {
    return this.isLocked;
  };

  this.setPassphrase = function(newPassphrase,
                                relockCallbackVoid) {
    return new Promise(function(resolve, reject) {
      if (this.isPassphraseSet() && this.isWalletLocked()) {
        reject("passphrase set/wallet is unlocked; can't set passphrase");
        return;
      }

      var success = function(response) {
        this.check = response.check;
        this.ephemeralKeyEncrypted = response.ekey_enc;
        this.salt = response.salt;
        this.setRelockTimeout(relockCallbackVoid);
        resolve();
      };

      var params = {
        'new_passphrase': newPassphrase
      };
      postRPC('set-passphrase', params).then(success.bind(this), reject);
    }.bind(this));
  };

  this.loadCredentials = function() {
    return new Promise(function(resolve, reject) {
      var params = {
        'check': this.check,
        'ekey_enc': this.ephemeralKeyEncrypted,
        'salt': this.salt
      };
      postRPC('set-credentials', params).then(resolve);
    }.bind(this));
  };

  this.clearRelockTimeout = function() {
    if (this.relockTimeoutId) {
      window.clearTimeout(this.relockTimeoutId);
      this.relockTimeoutId = undefined;
    }
  };

  this.setRelockTimeout = function(callbackVoid) {
    this.clearRelockTimeout();
    this.isLocked = false;
    this.relockTimeoutId = window.setTimeout(function() {
      this.lock(callbackVoid);
    }.bind(this), 1000 * 60 * 1);
  };

  this.lock = function() {
    return new Promise(function(resolve, reject) {
      this.isLocked = true;
      postRPC('lock', {}).then(resolve);
    }.bind(this));
  };

  this.unlock = function(passphrase, relockCallbackVoid) {
    return new Promise(function(resolve, reject) {
      postRPC('unlock',
              {'passphrase': passphrase}).then(function(response) {
                if (response.success) {
                  this.setRelockTimeout(relockCallbackVoid);
                  resolve();
                } else {
                  reject();
                }
              }.bind(this), reject);
    }.bind(this));
  };

  this.STORAGE_NAME = 'credentials';
  this.load = function() {
    return new Promise(function(resolve, reject) {
      var success = function(response) {
        if (response) {
          this.loadStorableObject(response).then(resolve);
        } else {
          this.init();
          resolve();
        }
      };
      var failure = function(response) {
        reject(response);
      };
      loadStorage(this.STORAGE_NAME).then(success.bind(this),
                                          failure.bind(this));
    }.bind(this));
  };

  this.save = function() {
    saveStorage(this.STORAGE_NAME, this.toStorableObject());
  };
}

File: nacl_module/credentials.cc
// Copyright 2014 Mike Tsao <mike@sowbug.com>

// Permission is hereby granted, free of charge, to any person
// obtaining a copy of this software and associated documentation
// files (the "Software"), to deal in the Software without
// restriction, including without limitation the rights to use, copy,
// modify, merge, publish, distribute, sublicense, and/or sell copies
// of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
// BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "credentials.h"

#include <iostream>  // cerr

#include "crypto.h"

const uint32_t KEY_SIZE = 32;

// echo -n "Happynine Copyright 2014 Mike Tsao." | sha256sum
const bytes_t PASSPHRASE_CHECK =
  unhexlify(
            "df3bc110ce022d64a20503502a9edfd8acda8a39868e5dff6601c0bb9b6f9cf9"
            );

Credentials::Credentials() {
}

Credentials& Credentials::GetSingleton() {
  static Credentials singleton;
  return singleton;
}

void Credentials::Load(const bytes_t& salt,
                       const bytes_t& check,
                       const bytes_t& encrypted_ephemeral_key) {
  salt_ = salt;
  check_ = check;
  encrypted_ephemeral_key_ = encrypted_ephemeral_key;
}

bool Credentials::SetPassphrase(const std::string& passphrase,
                                bytes_t& salt,
                                bytes_t& check,
                                bytes_t& encrypted_ephemeral_key) {
  if (isPassphraseSet()) {
    // We're changing an existing passphrase.
    if (isLocked()) {
      return false;
    }
  } else {
    // We're setting a new passphrase. Generate a new ephemeral key.
    ephemeral_key_.resize(KEY_SIZE);
    Crypto::GetRandomBytes(ephemeral_key_);
  }

  // Generate a new salt.
  salt.resize(KEY_SIZE);
  Crypto::GetRandomBytes(salt);

  // Generate the new key.
  bytes_t key(KEY_SIZE, 0);
  if (!Crypto::DeriveKey(passphrase, salt, key)) {
    return false;
  }

  // Determine check.
  check.resize(KEY_SIZE);
  if (!Crypto::Encrypt(key, PASSPHRASE_CHECK, check)) {
    return false;
  }

  // Encrypt the ephemeral key.
  if (!Crypto::Encrypt(key, ephemeral_key_, encrypted_ephemeral_key)) {
    return false;
  }

  // Save what we've just done.
  Load(salt, check, encrypted_ephemeral_key);

  return true;
}

bool Credentials::Unlock(const std::string& passphrase) {
  if (!isLocked()) {
    return false;  // in case someone uses result to verify a passphrase
  }

  // Generate the purported key.
  bytes_t key(KEY_SIZE, 0);
  if (!Crypto::DeriveKey(passphrase, salt_, key)) {
    return false;
  }

  // Does the check match?
  bytes_t check_decrypted;
  if (!Crypto::Decrypt(key, check_, check_decrypted)) {
    return false;
  }
  if (check_decrypted != PASSPHRASE_CHECK) {
    return false;
  }

  // Check is good. Decrypt the ephemeral key.
  if (!Crypto::Decrypt(key, encrypted_ephemeral_key_, ephemeral_key_)) {
    return false;
  }

  return true;
}

bool Credentials::Lock() {
  ephemeral_key_.clear();
  return true;
}
